couchdb/tests/server/compiler.py

```diff
@@ -99,6 +132,80 @@ def test_fail_for_relative_path_against_root_module(self):
                           './foo/bar/baz'.split('/'), {}, module)
 
+    def test_cache_bytecode_for_future_usage(self):
+        ddoc = {'foo': {'bar': {'baz': 'exports["answer"] = 42'}}}
+        require = compiler.require(ddoc)
+        exports = require('foo/bar/baz')
+        self.assertEqual(exports['answer'], 42)
+        self.assertTrue(isinstance(ddoc['foo']['bar']['baz'], types.CodeType))
+        exports = require('foo/bar/baz')
+        self.assertEqual(exports['answer'], 42)
+
+    def test_cache_egg_exports(self):
+        ddoc = {
+            'lib': {
+                'universe': DUMMY_EGG
+            }
+        }
+        require = compiler.require(ddoc, enable_eggs=True)
+        exports = require('lib/universe')
+        self.assertEqual(exports['universe'].question.get_answer(), 42)
+        self.assertTrue(isinstance(ddoc['lib']['universe'], dict))
+        exports = require('lib/universe')
+        self.assertEqual(exports['universe'].question.get_answer(), 42)
+
+    def test_reqire_egg_from_module(self):
+        ddoc = {
+            '_id': 'foo',
+            'lib': {
+                'egg': DUMMY_EGG,
+                'utils.py': (
+                    "universe = require('lib/egg')['universe'] \n"
+                    "exports['title'] = 'The Answer' \n"
+                    "exports['body'] = str(universe.question.get_answer())")
+            }
+        }
+        require = compiler.require(ddoc, enable_eggs=True)
+        exports = require('lib/utils.py')
+        result = ' - '.join([exports['title'], exports['body']])
+        self.assertEqual(result, 'The Answer - 42')
+
+    def test_required_modules_has_global_namespace_access(self):
+        ddoc = {
+            '_id': 'foo',
+            'lib': {
+                'egg': DUMMY_EGG,
+                'utils.py': (
+                    "import math\n"
+                    "def foo():\n"
+                    "  return math.cos(0)\n"
+                    "exports['foo'] = foo")
+            }
+        }
+        require = compiler.require(ddoc, enable_eggs=True)
+        exports = require('lib/utils.py')
+        self.assertEqual(exports['foo'](), 1)
+
+    def test_fail_on_resolving_deadlock(self):
+        ddoc = {
+            'lib': {
+                'stuff': (
+                    "exports['utils'] = require('./utils') \n"
+                    "exports['body'] = 'doc forever!'"),
+                'helper': (
+                    "exports['title'] = 'best ever' \n"
+                    "exports['body'] = require('./stuff')"),
+                'utils': (
+                    "def help():\n"
+                    "  return require('./helper') \n"
+                    "stuff = help()\n"
+                    "exports['title'] = stuff['title'] \n"
+                    "exports['body'] = stuff['body']")
+            }
+        }
+        require = compiler.require(ddoc)
+        self.assertRaises(exceptions.Error, require, 'lib/utils')
```

Comment:
It's a bit tricky, indeed: for the top-level design functions (map, show, vdu, etc) the path must be absolute. But CommonJS modules may use relative paths which makes them portable and reusable. Actually, there is no reason to use relative paths for design functions.

Today, I would indeed implement that feature through imports, but those days there was need to have 2.4 compatibility and all these sad things. Modern Python provides quite good import hooks to use. You could make pseudo module `ddoc` for all that stuff, for instance, and do all imports through it. However, that's a new feature and major improvement.